src/Npgsql/TopologyAwareDataSource.cs

```diff
@@ -234,46 +248,103 @@ private int getKeysByValue(CloudPlacement cp) {
         return GetPrivateOrPublicServers(currentPrivateIps, currentPublicIps);
     }
 
-    new List<string> GetPrivateOrPublicServers(List<string> privateHosts, List<string> publicHosts)
+    Dictionary<string, string> GetServerNodeTypeMap(Dictionary<string, string> serversNodeTypeMap)
     {
-        List<string> servers = base.GetPrivateOrPublicServers(privateHosts, publicHosts);
-        var flag = 0;
-
-        if (servers != null && servers.Any())
+        Dictionary<string,string> serversNodeTypeMapCopy = serversNodeTypeMap;
```

Comment:
Had to create a copy because the loop after this iterates over each entry of `serversNodeTypeMap` and removing any key during that loop from this results in error. So created a copy from which the keys can be removed and that copy is only returned.